Update iis.rst

GuardRex opened the pull request:
@danroth27 @Rick-Anderson @shirhatti @blowdart

Addresses https://github.com/aspnet/Docs/issues/1140

Dan: I set the base fork to `aspnet/Docs` and the base to `dev`. I hope that does the trick. If not, I'll have to get some more GH training from @PinpointTownes. His first course didn't cover this aspect.

There will be modifications, so I'll squash everything after we've made updates ... to the updates! :smile:

This addresses a number of areas, including the RC2 changes [EDIT: The prior PR that merely updates the name `5` to `Core` can be killed.]:

1. Order of presentation a bit more inline with the process one would follow
2. Expanded IIS setup for IIS Local with a screenshot
3. Removed content on unlocking handlers section of `web.config`
4. Added a special note on Nano Server to visit that doc for details
5. Updated references from ASP.NET 5 to ASP.NET Core
6. Updated references to the HTTP Platform Handler to the ASP.NET Core Module
7. Added information for installing .NET Core, with language and link on portable and self-contained app types WRT IIS hosting
8. Updated the deployment instructions, including new screenshots and a solid warning about the danger of an accidentally-misnamed or moved `web.config` file.
9. Rocked on the `Troubleshooting` section to reduce support requests ... trying to break $20B in operating income this year! :+1: [I hope Satya doesn't forget my cut! :smile:]
10. Added a couple of links to `Additional Resources`

https://github.com/GuardRex/Docs/blob/guardrex/publish-to-iis-rc2-update/aspnet/publishing/iis.rst

dnfclas commented:
Hi __@GuardRex__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

PinpointTownes commented:
@GuardRex you need to rebase your PR: run `git fetch upstream`, `git rebase upstream/dev`. Fix the conflicts, run `git add .`, `git rebase --continue` and then `git push -f` :smile:

dnfclas commented:
Hi __@GuardRex__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

dnfclas commented:
Hi __@GuardRex__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

GuardRex commented:
@danroth27 ... sorry about that. The idiot :monkey: had a problem over here; but after banging on the keyboard, apparently solved it. I think it's ok now.

After additional updates, we'll see if the :monkey: knows how to rebase without draining @PinpointTownes cell phone battery! :smile:

shirhatti commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
+=================
 
-By `Rick Anderson`_ and `Luke Latham`_
+By `Rick Anderson`_ and `Luke Latham <https://github.com/GuardRex>`_
```
If you want to add a link to your name, I'll suggest doing it in the [`authors.txt` file](https://github.com/aspnet/Docs/blob/master/common/authors.txt) instead

shirhatti commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
 
-Configure Data Protection
------------------------------------
+IIS Configuration
```
IIS Configuration content should not live on the ASP.NET Core Docs.
That content will live on the IIS docs. I would recommend dropping that content from this article like in the rc1 version of this doc

GuardRex replied:
Ah! This is one that I hope you go in one of two directions with to address one of my pet peeves about MS documentation over the last ... well ... since 1975! :smile: This doc isn't meant to be a "walkthrough" or "tutorial" as many MS docs are not meant to be walkthroughs or tutorials. **However,** that's exactly what you have with devs coming into complex new technologies (new to them notwithstanding the new-ness of .NET Core). They're looking for a fairly comprehensive end-to-end document that covers all of the basic config and that walks them through to a successfully running app.

Further, my experience with devs is that they like a few of the more notable gocthas, warnings, tips, and troublshooting points. They all love :heart:, as I do, a comprehensive and well-organized Troubleshooting section that covers all of the basic mis-steps and mis-configs. I hope you'll go in either of these two directions with this ...

1. Leave this doc with the added bits. Don't split IIS config out. Don't hack a mere handful of comman IIS goofs from the list. Let it be the "Go To" doc for publishing .NET Core on IIS. I was only half-joking when I said MS can hit $20B this year with this kind of approach: Every support contact is at least $7 (a typical industry figure that was given to my the VP of an international telecom). Keeping this doc all together as I show will very likely minimize support requests, thus save MS operating expendatures.
2. On the other hand, Yes!, indeed, get the IIS config out (at least) ... but ... also provide a proper `ASP.NET Core on IIS` tutorial. That would perhaps be even better, because we could get into `project.json` config, attributes on the `<aspNetCore>` element, have all of the Troubleshooting (including a handful of purely IIS Common Issues that devs come up against), and more "gotchas," warnings, and tips with more IIS config commentary (if desired).

Ok ... TL;DR ... I hope you'll either leave this doc as is or agree that a full tutorial is warranted. I volunteer to write it. Please don't strip this doc of IIS config and IIS-only Troubleshooting without a tutorial. MS will lose money, and I et al. will end up wasting a lot of time on JabbR helping devs with basic IIS hosting config and troublshooting. Thanks for hearing me out.

shirhatti replied:
> I hope you'll either leave this doc as is or agree that a full tutorial is warranted. I volunteer to write it.

A full tutorial is warranted, this is just not the destination for it.

> They're looking for a fairly comprehensive end-to-end document that covers all of the basic config and that walks them through to a successfully running app.

The top-level node [Hosting](https://docs.asp.net/en/latest/hosting/index.html) is meant to be more comprehensive and I definitely see how we can shoehorn some of this great content you've written to live over there. Though not it's current shape, it should eventually read as a whitepaper for anybody hosting an ASP.NET Core application. I definitely think some of this installing IIS and troubleshooting content can live there. In fact, I would love to eventually see some more content there w.r.t. Recycling settings on AppPools, SSL Cert binding, etc., all of which are pertinent to hosting an ASP.NET Core application.

@danroth27 Do you agree? tl;dr prune this article, find a new home for some of the content

GuardRex replied:
> this is just not the destination for it.

It would be in-line with other existing tutorials, such as `ASP.NET 5 on Nano Server` and `ASP.NET 5 and Azure Service Fabric`. It would also be easy to find if its over here.

Aequoris replied:
I found GuardRex's article helpful in its current form. I have been building prototype apps for startups using ASP.NET Core, and delivering it to dev teams with no familiarity with ASP.NET Core, to run and extend from that point on.

Having all the troubleshooting steps in a single place, has already enabled a team to handle the deployments themselves, any errors that were encountered, were solved with that document.

Splitting it up, would make it harder to find the answers to the problems.

danroth27 replied:
The problem is that we don't want to maintain docs for technologies that we don't directly own. What happens if the IIS team makes some changes that affect these troubleshooting steps? These docs are the ASP.NET Core docs, not the IIS docs. That doesn't mean we can't link to the IIS docs liberally, but we need to make sure we document ASP.NET Core before we try to document other technologies. We have plenty that we still need to document for ASP.NET Core.

GuardRex replied:
[What follows is a little off-topic ... and a **BOOK**, but humor me for a minute. I just want to get my thoughts down and out on this. As an educator, this is important to me. Read the TL;DR and skip it. I just want to say my peace and rest in peace :skull:. I know a dude ... a dude who was big at MS back in the '90's ... he is good friends with Satya ... yeah ... "a dude." He just told me the other day that MS has the same challenges that all giant corporations have. Things like politics, difficulty in communication, difficulty in executing large-scale strategies, and difficulty holding people accountable are all alive and well there to spite best efforts. It's easy for me to sit out here and shoot my mouth off. I get it. However, you did OSS this. So ... read the TL;DR and skip the rest. I'm going on the record though with **one** of my pet peeves about MS docs across the company.]

TL;DR ... I'll chop the IIS Config here and chop the purely IIS Troubleshooting bits; however, please consider a full tutorial `ASP.NET Core on IIS` that keeps **ALL** of this stuff in one place and adds a little more **cowbell** :bell:.

If not, ok then ... @benaadams just suggested to me ... blog it and those of us on the "JabbR Outreach User Support Team" (JOUST in superhero graphic novel circles) can point to the GuardRex blog. We can do that. That sucks ... not what I want to do ... but MS is MS and GuardRex is GuardRex, and I know what I can control. :smile:

> we don't want to maintain docs for technologies that we don't directly own.

... which I take as a flaw in the organization of Microsoft business units, philosophy, and effort to find a solution. I doubt that this is an intractable problem.

I have a very strong suspicion that if we divided up a large group of devs randomly into two groups and gave one group this doc in its present form and gave the other group this doc with the IIS bits (IIS config and purely IIS troubleshooting cut out), that the first group would :guitar: **ROCK** :guitar:. <---- That should be the guide ... what works for the **customer** ... not what is expedient or simple for the company. I suspect MS will lose far more money in operational support with crappy docs than it will by investing in a solution to this problem.

Sure, IIS can throw many more exceptions than what I show here. I was just hitting the common ones that I've seen devs messing up in the trenches over the past year. If there is a great IIS doc that runs down all of the basic IIS screw-ups a dev can make and what to do about them, that doc should be linked here with no mention of those problems here. One little problem though ... that doc **doesn't exist** AFAIK. You want to link this??? ... http://www.iis.net/learn/troubleshoot ??? Yuck. Which IIS doc covers the minimal IIS config for .NET Core app hosting on WS IIS and local IIS? Which one? ... Yeah ... It doesn't exist. It's only here in this revision.

It's inefficient to create unnatural breaks in docs because one team "owns" a piece of a process and some other team "owns" another piece. There has to be an approach/philosophy/process/engineering solution that would allow the IIS folks to "own" and be responsible for their pieces of this doc (the purely config and troubleshooting bits) while leaving all of this content accessible here, which is a one-stop-shop for ASP.NET Core on IIS with the most common dev screw-ups based on working with devs in the field.

If this doc is cut without a comprehensive tutorial, then devs will probably suffer. MS will probably lose money. .NET Core on IIS will be a sour first experience for many. I'm telling you from the trenches on JabbR that I have been in conversations with frustrated devs, and I've seen them walking away from things when the docs (because we're in RC) don't support what they are trying to implement.

I've seen @davidfowl spend hours re-explaining things, including IIS config. One of his recent conversations re-explaining something was 1.5 hours. I timed it. We were laughing about it afterwards to reduce the pain. The current version of the IIS revision is getting hits right now. It's actually (as @Aequoris just said) useful in its current comprehensive form. He's not the only one: Two more devs on JabbR said they fixed their IIS hosting probs based on this doc revision. This has nothing to do with .NET Core in RC2 [or RC17 (haha)] ... if this doc in its present form saves @davidfowl even 10 minutes, it was worth it leaving the IIS bits in it.

The problem here is the same old problem truly going back to 1975! MS engineers are great at building things and they often suck at explaining to end-users how to use things. Documentation is still far-too-frequently an after-thought by engineers and teams. Oftentimes, the team just doesn't have the communications "chops," willpower, or available time to do the job properly.

Take the Nano Server Team for example: They are **insanely GREAT engineers** ... I :heart: those :cat:s and Nano is **amazing technology**, but their docs are insanely **bad**. Why? Plain and simple, they're just not that good at explaining things to people. I guess that like many high-end engineers that they went into technology to build things and not to teach things. The lack of experience in education and communication combined with little desire to sit for many hours working on a single doc is obvious. I spent four full days on this revision. Can you really see any of them with their interest in engineering and their backlog being able to do that? No way. That's not really a knock on them. I think this is a fact of business unit organization there very heavily ... too heavily as far as docs are concerned ... centered on engineering. I think the problem is actually a lack of creative thinking on how to support documentation in a giant organization.

Maybe Apple has figured it out ... maybe Adobe has figured it out. idk ... but Microsoft surely hasn't figured it out, or else we wouldn't be talking about cutting **critical** parts of this doc without addressing the developer pain that will ensue. We wouldn't see MS using at least a dozen completely different doc platforms across the company that have no other connection than web links (and they're poorly linked at that). [Ok ... ok ... I'm being a little melodramatic here! :smile: ... one gets so few chances to wax poetic in tech circles these days. lol]

Put the same creative juice that goes into MS products and services into docs. What is needed is some lateral thinking here. The model of each technology unit "owning" its own docs and never talking about another unit's technology is clearly not working for many types of cross-technology-cutting docs, introductory docs, tutorials, and samples.

... and they lived happily ever after. THE END

shirhatti commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
-If you need to enable IIS, see `IIS server configuration`_.
+\*Conceptually, the IIS configuration described in this document also applies to hosting ASP.NET Core applications on Nano Server IIS, but refer to `ASP.NET Core on Nano Server <https://docs.asp.net/en/latest/tutorials/nano-server.html>`_ for specific instructions.
 
-Configure Data Protection
```
I'm pretty, we still need to configure data protection on a per AppPool basis. You cannot drop this section
@danroth27 @pakrym can confirm this.

GuardRex replied:
... and a good spot to mention in passing that I thought the process follows this sequence (and getting the heading levels striaghtened out) (and subject to change based on your final decisions on section further discussed below) ... 

| Content | Level | Markup |
| ------- | ----- | ------ |
| Title: Publishing to IIS | Title | H1 as = |
| .... Supported Operating Systems | Section | H2 as - |
| ........  Windows Desktop Operating Systems | Subsection | H3 as ^ |
| ........  Windows Server Operating Systems | Subsection | H3 as ^ |
| .... IIS Configuration | Section | H2 as - |
| .... Install .NET Core on the Server | Section | H2 as - |
| .... Deploy the Application | Section | H2 as - |
| .... Configure the Website in IIS | Section | H2 as - |
| .... Create a Data Protection Registry Hive | Section | H2 as - |
| .... Common Errors | Section | H2 as - |
| .... Additional Resources | Section | H2 as - |

shirhatti commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
 
-Deploy to IIS server
------------------------------------
+Install the Microsoft ASP.NET Core Module version 1.0 or higher on the server:
```
The rc2 version of ASP.NET Core Module will not be available as a standalone .msi.

The recommended way of acquiring ANCM is installing ASP.NET Core 1.0 RC2 server bundle that contains both the .NET Redist and ANCM.

When the download link becomes live, we will want to point to that bundle.

GuardRex replied:
Ok ... I'll fix this. Do you know if the bundle will be called, as it is today, the `.NET Core Installer`?

danroth27 replied:
Looks like it's going to be called the ".NET Core Windows Server Hosting" bundle.

shirhatti commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
+- `AspNetCoreModule x86 MSI Installer <http://download.microsoft.com/download/7/2/D/72DD68D5-7386-4A41-8B8F-E21779A69E33/aspnetcoremodule_x86_en.msi>`_
 
-.. image:: pubIIS/_static/b8.png
+Install .NET Core on the Server
```
Like I mentioned in my previous comments, you won't have to install the shared framework and ANCM independently. You will just be installing the bundle.

shirhatti commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
 
-IIS server configuration
------------------------------------
+If you are planning to host all of your applications as self-contained applications on your server (i.e., you will not deploy portable applications that require installing .NET Core on the server), you may skip this step. For more information on the difference between portable applications and self-contained applications, see `Types of portability in .NET Core <https://github.com/blackdwarf/core- ...
```
Even if we do away with section, it's worth keeping this note.
I would point to the published doc at http://dotnet.github.io/docs/core-concepts/app-types rather than the source of the doc

GuardRex replied:
However, we can't keep this blurb if the only way to get ANCM is via the bundle. They will be installing both ... it's just that when they load their self-contained apps calling their included executables, the shared host will not be triggered. Therefore, we can kill this whole bit.

Side Note: I really don't like this: I definitely **do not** want the shared framework on my production servers. Please I hope you are pushing forward on OSS'ing the ANCM. Those of us who feel as I should be able to compile our own ANCM and put that on our servers independently of the bundle for our self-contained-only apps scenarios.

danroth27 replied:
@GuardRex Can you tell us more about why you don't want to deploy .NET Core on your servers? You can a bunch of benefits - your apps will be much smaller and quicker to deploy, the installed framework will have native images already, etc.

GuardRex replied:
- Total app isolation (better fault-tolerance IMO; possibly better security model)
- No frameworks to maintain on the servers (minimal server config; minimal maintenance)
- Different apps on different runtimes
- No auto-opt-in to the telemetry program without having to maintain env vars on servers. Running apps as executables means no `dotnet` commands running on servers, which means no telemetry to MS. Besides, I hate env vars on servers anyway ... i.e., "NUTJOB" apps configure themselves https://gist.github.com/GuardRex/f40f1b1414e175a6514f

shirhatti commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
+
+.. include:: ./dataProtectionWarning.txt
+
+Common Errors
```
@danroth27 Do you want to surface this content here or someplace else? My only concern it almost doubles the length of the article giving the incorrect impression that publishing to IIS is difficult.

shirhatti replied:
One option is to prune the error list here and surface an exhaustive list someplace else.

danroth27 replied:
I'm fine with having a troubleshooting section at the end of this doc as long as the issues are ASP.NET Core specific. I don't want this to turn into a general IIS troubleshooting section. More general troubleshooting tips should live on iis.net, and then we can link to them from here.

GuardRex replied:
Thank the :pray: server gods :pray: for that!

Yes, picking up on my point above: Please save dev :broken_heart:'s, chatroom pros helping new and junior devs ... :boy: :gun:, and MS some serious support :moneybag: money :moneybag:. I'm speaking from quite a bit of experience helping folks with getting their apps going on IIS. This expanded and comprehensive IIS Troubleshooting section is just what the doctor ordered :pill: ... IF ... IF ... we don't see a real tutorial up there that would have this full section.

GuardRex replied:
Will prune the purely IIS ones ... **Arrrrrrrrrrrrrr** :-1: but your call, of course.

danroth27 commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
-.. contents:: Sections:
-  :local:
-  :depth: 1
+- `Supported Operating Systems`_
```
Please use the `contents` directive instead of explicitly listing out links to section headings.

danroth27 commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
-Install the HTTP Platform Handler
------------------------------------
+Supported Operating Systems
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Use `-` for H1, `^` for H2.

shirhatti commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
+Visit the `GitHub dotnet/cli repo (rel/1.0.0 branch) <https://github.com/dotnet/cli/tree/rel/1.0.0>`_ to obtain the correct MSI installer or binaries. To install .NET Core using the MSI, obtain the correct **.NET Core Installer** for your server. If you prefer to install from binaries, obtain the correct **.NET Core Binaries** for your server. These selections will only install .NET Core assembli ...
 
-.. image:: pubIIS/_static/rs.png
+.. note:: If you install .NET Core from binaries, either include the path to `dotnet.exe` in the system PATH environment variable or make sure that your `web.config` file `\<aspNetCore\>` element `processPath` attribute includes the full path to `dotnet.exe` on the server.
```
Shouldn't be required any longer. The installer bundle puts `dotnet.exe` on your path.
However, you have to run `iisreset` or reboot your machine for IIS to pick up changes to your system path

GuardRex replied:
The reason I mentioned the PATH was that I thought you were still going to make the binaries avaiable. If I'm hearing you correctly, the binaries for the shared framework are going away in favor of the bundle. Therefore, I will remove the bit about the PATH.

I will add here that `iisreset` must be issued or the server should be restarted after the bundle is installed.

shirhatti commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
+^^^^^^^^^^^^^^^^^^^^^^
 
-.. image:: pubIIS/_static/role-services.png
+#. On the target IIS server, create a folder to contain the application's assets.
```
I'd recommend creating a directory for the log files as well. My guidance would be to create with the same path that the `publish-iis` tool generates by default since ANCM cannot write logs to disk if that directory doesn't exist.

GuardRex replied:
> ANCM cannot write logs to disk if that directory doesn't exist

[Side Note: There is also an open issue on having to have a dummy file in a `logs` folder in the project to get `dotnet publish` to get the folder into published output when it's listed in `publishOptions > include`. That's scheduled for RTM.]

When I wrote "application's assets" here, I was referring to the contents of the `publish` folder created by `dotnet publish`, including things like `wwwroot`, `logs`, `Views`, and perhaps `web.config` if they don't use @moozzyk's tooling. I have my apps setup with the tooling, and I also have a logs folder in `publishOptions`. I like to get the whole bleed'in thing in one spot (`publish`) and PowerShell the bits into place on the servers.

Two options here: (1) Assume they don't know about creating a `logs` folder in the app and having it in `publishOptions > include` and instruct them to do that, or (2) Take the approach that they shouldn't create the `logs` folder in the app and just tell them create the `logs` folder right here as soon as they create the app assets folder I'm calling for. You're suggesting the latter approach?

GuardRex replied:
I'll take the "latter approach" that matches what you said @shirhatti

shirhatti commented:
@GuardRex Great job on the article

GuardRex commented:
@shirhatti @danroth27 Thanks ... and thanks for the corrections. There is only one that trips me up a bit dealing with splitting the doc. I'll explain my concern more in the line notes on that one.

I'll make a pass through your comments by the end of day and clarify the changes that need to be made. I just want to make sure I understand clearly before I sit down to modify the language.

No rush on getting back to me. I'm just working my job search for the next few days through the weekend and playing with Nano TP5 .. fun FUN! :smile: :smile: :smile:

danroth27 commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
+- Windows Server 2008 R2 and newer\*
 
-If you need to enable IIS, see `IIS server configuration`_.
+\*Conceptually, the IIS configuration described in this document also applies to hosting ASP.NET Core applications on Nano Server IIS, but refer to `ASP.NET Core on Nano Server <https://docs.asp.net/en/latest/tutorials/nano-server.html>`_ for specific instructions.
```
Make all inline hyperlinks anonymous by following with two underscores: `` `blah <http://foo>__` ``

GuardRex replied:
Two questions ...
1. That should be \`blah <http://foo>\`__ ... correct?
2. all inline hyperlinks? http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#anonymous-hyperlinks ... I'm confused. The descriptions on how and why they exist seems at odds with doing it everywhere. Just to be clear: You want **all** inline hyperlinks with the double-underscore? 

btw- That aspect of links is not in the style documentation.

danroth27 replied:
> That should be `blah http://foo`__ ... correct?

Oops, fixed above.

> You want all inline hyperlinks with the double-underscore? 

Yes. The problem with one underscore is that you are implicitly defining a link target for the friendly text of the link. So `` `foo <http://bar>`_ `` is really the same as:

``` rst
foo_

.. _foo: http://bar
```

For one link this isn't an issue, but if you will get a warning if you try to define two links with the same friendly text like this:

``` rst
See `here <http://microsoft.com>`_.
Now see `here <http://google.com>`_.
```

Built that and you get the warning: `Duplicate explicit target name: "here".` 

If you instead use a double-underscore then no reusable target is defined and you get a "one-off" hyperlink, which is typically what most people intended. See http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#embedded-uris-and-aliases for more details.

You are right though that we should get this clarified in our style guide. Really we should be using explicit link targets instead of in-lining URIs everywhere so that the raw text is easier to read, but for now I just want to avoid noisy build warnings.

danroth27 commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
+
+A quick way to determine if the IIS reverse proxy to the Kestrel server is working properly is to perform a simple static file request (e.g., a stylesheet, script, or image) from the application's static assets (i.e., those in `wwwroot`) using `Static File Middleware <https://docs.asp.net/en/latest/fundamentals/static-files.html>`_. If the application can serve static files but MVC Views and othe ...
+
+In the following list, **ANCM Log** refers to the **ASP.NET Core Module** log and will be found on the path provided in the `stdoutLogFile` attribute of the `\<aspNetCore\>` element in `web.config`.
```
I think it reads better to just inline the module name instead of using the ANCM acronym.

danroth27 commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,187 @@
+^^^^^^^^^^^^^^^^^^^^
 
-- HTTP 500.21 Internal Server Error.
+- `Understanding ASP.NET 5 Web Apps <http://docs.asp.net/en/latest/conceptual-overview/understanding-aspnet5-apps.html>`_
```
ASP.NET Core

GuardRex replied:
This was just because the current (RC1) doc is titled this way at that link.

I'll go "Core" and guess that the doc location will be:
```
http://docs.asp.net/en/latest/conceptual-overview/understanding-aspnetcore-apps.html
```

danroth27 replied:
Right, the doc location won't change. It's already been updated with the new name in the dev branch, which will get merged to be "latest" once we ship RC2.

GuardRex replied:
Ah, yes. That was dumb of me not to check the `dev` branch. Gotcha!

shirhatti commented:
@GuardRex I think Dan captured my concern perfectly. I wasn't advocating splitting up the doc, but rather limiting that list to ASP.NET Core specific errors. For e.g., Application Pool not started is not an error that should live on this list.

GuardRex commented:
@shirhatti I have a suggestion coming up on that. What is really needed IMO that would allow the IIS config to go away + more tips, gotchas, warnings, app config (i.e., `project.json` & `WebHostBuilder`), ANCM details, and a full Troubleshooting section ... **IS** (drum-roll please!) ... an `ASP.NET Core on IIS` **tutorial**. :clap: :clap: See my point? Devs really need to be walked through it, and it would clean those bits out of the doc that aren't strictly `Publishing to IIS`. I'd volunteer to write that one btw.

On the other hand though, I'm going to make the case to **PLEASE** don't gut this doc with so many devs looking for a one-stop-shop on how to get .NET Core apps on IIS if there isn't going to be a tutorial. I really think this is a major pain-point for MS docs over the last few **decades**.

I have a counter-argument to your comment that I've already typed out, and I'm about to just run down the comments and paste in my questions and comments. I'll have it up in a few minutes. All I ask is hear me out on it before you cats decide.

GuardRex commented:
@danroth27 @shirhatti Ok ... your cell phones might have just gone nuts. I've put my questions and comments there. No rush on getting back ... but I'm going to hold-off on working it forward until I'm clear on your guidance.

GuardRex commented:
Last Note: I have one last failure scenario for `Troubleshooting` to look at: I discovered ...

1. Add `"platform": "x86"` to `buildOptions`
2. Set the RID to `win81-x86`
3. Restore
4. `dotnet publish -r win81-x64` ... It will publish!
5. Deploy
6. :boom:

... apparently because the RID for restore and publish were not the same. I need to repro that mismatch again and study the IIS behavior. If it sticks, I'll add it to the list and ping you.

moozzyk commented:
@GuardRex - what kind of an app is this (portable vs standalone)? I found https://github.com/dotnet/cli/issues/2884 today...

GuardRex commented:
@moozzyk Because I know the test was done with an RID specified in the `dotnet publish -c Release -r win81-x64`, I know that happened with my self-contained test app.

> I found dotnet/cli#2884 today

Yes! Indeed. The only thing that you didn't report is an outcome similar to what I found when I actually deployed those bits ... it went :boom:. I'll test this scenario soon and make sure I know the repro scenario and browser/log result.

moozzyk commented:
@GuardRex - I think @anurse looked a while ago at 32-bit processes. Not sure what the outcome was - I think it is this:  https://github.com/dotnet/cli/issues/2428

anurse commented:
Yeah, unfortunately for desktop this is a little gross since we don't currently build separately for each RID, we just compile once and copy stuff around. On Desktop, you need to set the `/platform` switch to ensure the process is loaded with the correct bitness. We have thoughts on how to improve this, but for now you have to go through some weirdness to make it work. If you need to build for both x86 an x64, you can use `configurations` for now, to work around the issue: https://github.com/dotnet/coreclr.xunit/pull/28/files#diff-54c2b73d20330ac1306b439ffc7070bcR16

GuardRex commented:
I just verified the repro for a missing warning message or missing error message:

1. Add `"platform": "x86"` to `buildOptions`
2. Have the RID to an `x64` flavor RID ... e.g., `win81-x64`
3. Restore
4. `dotnet publish -r win81-x64` ... It will publish fine with no warnings or errors
5. Deploy
6. :boom:

The result is ...
Browser:
```
No response until the request times out.
```
ANCM Log: 
```
Unhandled Exception: System.BadImageFormatException: Could not load file or assembly 
'teststandalone.dll' or one of its dependencies. An attempt was made to load a program with 
an incorrect format.
```
Application log (via Event Viewer): 
```
Faulting application name: teststandalone.exe, version: 0.0.0.0, time stamp: 0x572a8e4a
Faulting module name: KERNELBASE.dll, version: 10.0.10586.162, time stamp: 0x56cd45b4
Exception code: 0xe0434352
Fault offset: 0x0000000000071f28
Faulting process id: 0xce4
Faulting application start time: 0x01d1a7acfa2b3ce2
Faulting application path: C:\_APPLICATIONS\teststandalone\bin\Release\netcoreapp1.0\win81-x64\publish\teststandalone.exe
Faulting module path: C:\WINDOWS\system32\KERNELBASE.dll
Report Id: a413017d-cc75-4899-a1c9-00773bfc90a8
Faulting package full name: 
Faulting package-relative application ID: 
```

I don't think this repro quite matches any of your, @moozzyk, open issues, does it? If not, @anurse, should I open it in `dotnet/cli`? ... seems like it should be detected and fail at publish/compile time.

Even if it is opened for a warning or error message, this will not make the cut for RC2, correct? Therefore, should it go into the `Troubleshooting` section here? I can see a few devs accidentally making this mistake. We can cut it from `Troubleshooting` later (RTM or next release). /cc @shirhatti @danroth27 @Rick-Anderson

moozzyk commented:
@GuardRex - please file it in the dotnet/cli repo.

GuardRex commented:
@danroth27 @Rick-Anderson @shirhatti 

I think I hit everything. Ready for 2nd draft review.

- It will need the link to the bundle.
- @blowdart ... I enhanced the Warning in the Deploy the Application section. I try to be very clear, and I added a "gotcha" bit at the end about disabling entire file extensions because it will break legitimate static file serving out of `wwwroot`.
- I added a wee bit more to `Troubleshooting` making it utterly clear how to enable application logging ... noting that the `publish-iis` tooling disables it by default. I note the issue about making sure the full path for saving logs is present. I add a line about enabling application logging for exceptions that are generated within the app. I'm trying to encourage devs not to just show up on JabbR with unknown exceptions within the app (e.g., 500, routing issues) where we ask 'Did you enable logging?' and they say that they didn't. Maybe a waste of time, but if it saves @davidfowl ten minutes, I think it's worth the added verbage here.
- I added the faulting module exception where `platform` is set to a different bitness than the publish RID for self-contained apps.

GuardRex commented:
For reference on platform bitness not matching self-contained publish RID: https://github.com/dotnet/cli/issues/2922

shirhatti commented:
> It will need the link to the bundle.

We don't have a link ready just yet. In the meanwhile, I'd just point to https://microsoft.com/net and create a separate issue to track updating all the download links

shirhatti commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,178 @@
-#. Navigate to the publish folder (*..\\..\\artifacts\\bin\\WebApp9\\Release\\Publish folder* in this sample).
-#. Copy the **approot** and **wwwroot** directories to the target IIS server. Note: MSDeploy is the recommended mechanism for deployment, but you can use Xcopy, Robocopy or another approach. For information on using `Web Deploy` see :doc:`iis-with-msdeploy`.
-#. In IIS Manager, create a new web site and set the physical path to **wwwroot**. You can click on **Browse *.80(http)** to see your deployed app in the browser.
+#. Install the `.NET Core Windows Server Hosting <http://download.microsoft.com/download/7/2/D/72DD68D5-7386-4A41-8B8F-E21779A69E33/aspnetcoremodule_x64_en.msi>`__ **[UPDATE LINK TO BUNDLE LOCATION]** bundle on the server. The bundle will install .NET Core assemblies and the **ASP.NET Core Module**. The module creates the reverse-proxy between IIS and the Kestrel server.
```
> .NET Core assemblies

nit: .NET Core Runtime

nit: **ASP.NET Core Module** -> ASP.NET Core Module

shirhatti commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,178 @@
+----------------------------
 
-3. Unlock the configuration section.
+#. In **IIS Manager**, create a new website. Provide a **Site name** and set the **Physical path** to the application's assets folder that you created. Provide the **Binding** configuration and create the website. [See the screenshots below.]
```
nit: [See the screenshots below.] is unnecessary

shirhatti commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,178 @@
 
-3. Unlock the configuration section.
+#. In **IIS Manager**, create a new website. Provide a **Site name** and set the **Physical path** to the application's assets folder that you created. Provide the **Binding** configuration and create the website. [See the screenshots below.]
+#. Set the application pool to **No Managed Code**. [See the screenshot below.] ASP.NET Core runs in a separate process and manages the runtime.
```
nit: [See the screenshots below.] is unnecessary

shirhatti commented:
@danroth27 :shipit: from me besides a couple of nits. Can we have a UE review as well?

GuardRex commented:
@shirhatti @danroth27 @Rick-Anderson 

Updates made and issue opened for bundle link update: https://github.com/aspnet/Docs/issues/1192

When you pull the trigger for a possible rebase, do you want me to ...

1. rebase on :shipit:, because this PR will get merged and the bundle link will be a separate PR, or 
2. rebase later, because the bundle link can be provided here for this PR merge?

shirhatti commented:
Rebase now and merge.

We'll sweep the entire repo and updates any download link changes (not just this article).

GuardRex commented:
@shirhatti Maybe that worked. There was a little hand-to-hand combat with git on the rebasing, so could you check and let me know if this is ok?

shirhatti commented:
> There was a little hand-to-hand combat with git on the rebasing, so could you check and let me know if this is ok?

Rebasing looks good

shirhatti commented:
@danroth27 **ping**
Can we have someone do a UE review before I merge this?

Rick-Anderson commented:
@GuardRex @shirhatti @danroth27  I'll do it tonight

Rick-Anderson commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,178 @@
-=============================
+=================
 
 By `Rick Anderson`_ and `Luke Latham`_
```
Change order 
By Luke and Rick

GuardRex replied:
Thanks ... and bound to be helpful with potential employers. :smile:

Rick-Anderson commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,178 @@
------------------------------------
+#. On the target IIS server, create a folder to contain the application's assets.
+#. Within the folder you created, create a **logs** folder to hold application logs (if you enable logging). If plan you deploy your application with a **logs** folder in the payload, you may skip this step.
+#. Deploy the application to the folder you created on the target IIS server. MSDeploy (Web Deploy) is the recommended mechanism for deployment, but you may use any of several methods to move the application to the server (e.g., Xcopy, Robocopy, PowerShell). Visual Studio users may use the `default Visual Studio web publish script <https://github.com/aspnet/vsweb-publish/blob/master/samples/defau ...
```
replace e.g. with for example

Rick-Anderson commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,178 @@
 
-1. Enable the **Web Server (IIS)** server role. In client operating systems (Windows 7 through Windows 10) select **Control Panel > Programs > Programs and Features > Turn Windows features on or off**, and then select **Internet Information Services**.
+.. warning:: 
+  .NET Core applications are hosted via a reverse-proxy between IIS and the Kestrel server. In order to create the reverse-proxy, the `web.config` file must be present at the content root path (i.e., the application base path) of the deployed application, which is the website physical path provided to IIS.
```
replace i.e. with "that is" or something like
file must be present at the content root path (the app base path) of the 
or
file must be present at the content root path (typically the app base path) of the

Rick-Anderson commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,178 @@
+  .NET Core applications are hosted via a reverse-proxy between IIS and the Kestrel server. In order to create the reverse-proxy, the `web.config` file must be present at the content root path (i.e., the application base path) of the deployed application, which is the website physical path provided to IIS.
 
-.. image:: pubIIS/_static/rs.png
+  On the application's physical path, there are sensitive files that one would never want to accidently serve, such as **my_application.runtimeconfig.json**, **my_application.xml** (XML Documentation comments), and **my_application.deps.json**. Without the `web.config` file present to create the reverse proxy to Kestrel, IIS would serve these sensitive files openly on the Internet. **Therefore, i ...
```
delete "openly on the Internet"

Rick-Anderson replied:
Consider rewriting as 
There are sensitive files on the app's physical path such as **my_application.runtimeconfig.json**, **my_application.xml** (XML Documentation comments), and **my_application.deps.json**.  The *web.config* file is required to create the reverse proxy to Kestrel, preventing these files from being served.

GuardRex replied:
Changing slightly from what you show to avoid state verbs "are" and "being" ...

> Sensitive files exist on the app's physical path, such as `my_application.runtimeconfig.json`, `my_application.xml` (XML Documentation comments), and `my_application.deps.json`. The `web.config` file is required to create the reverse proxy to Kestrel, which prevents IIS from serving these files.

Rick-Anderson replied:
@GuardRex OK

Rick-Anderson commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,178 @@
+  In order to mitigate the risk of serving sensitive application files if `web.config` were ever accidently renamed or removed, one may add sensitive files to the **Hidden Segments** of **Request Filtering** or add blanket file extension rules to the **File Name Extensions** of **Request Filtering**. See `Hidden Segments \<hidden Segments\> <https://www.iis.net/configreference/system.webserver/se ...
 
-.. image:: pubIIS/_static/role-services.png
+  Keep in mind that if you use the **File Name Extensions** approach that your application will not receive requests nor be able to serve any file with that extension, even static files in your **webroot** that you wish to serve (e.g., JSON or XML files you have placed in `wwwroot` for legitimate file serving). Therefore, specifically naming sensitive files using the **Hidden Segments** approach  ...
```
replace all e.g. with "for example"
replace all paths/files with *italic* and remove `code` backticks. I won't make that comment again.

Rick-Anderson replied:
>might be preferable in most scenarios
too many qualifiers.
How about "may be preferable"

Drop the [ and just make it a new sentence.

Rick-Anderson commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,178 @@
-  - Launch IIS Manager and select the server in the **Connections** pane on the left (see image below).
-  - Double-click **Configuration Editor**.
-  - In the **Section** drop-down, select **system.webServer/handlers**, and then click **Unlock Section**.
+.. note:: If you change the default identity of the application pool from **ApplicationPoolIdentity** to some other user account, make sure you configure the correct permissions for the user identity to use the application's assets folder.
```
drop "to some other user account"
consider
If you change the default identity of the application pool from **ApplicationPoolIdentity**, verify the new identity has the required permissions to access the application assets an database.

Rick-Anderson commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,178 @@
 
-Supported operating systems
------------------------------------
+	.. image:: pubIIS/_static/editapppool.png
```
@shirhatti don't we want to show browse web site (show them how to browse site) at this point?

shirhatti replied:
If you think it would read better, I'm all for it.

GuardRex replied:
I'll add it.

Rick-Anderson commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,178 @@
+	.. image:: pubIIS/_static/editapppool.png
 
-The following operations systems are supported:
+Create a Data Protection Registry Hive
```
@shirhatti  - shouldn't this be titled "Recommended"?

shirhatti replied:
Huh? I don't follow.

Create a Data Protection Registry Hive is more 'mandatory' than 'recommended', but maybe I completely missed the point you were trying to make.

Rick-Anderson replied:
OK, got it. Leave it as is.

Rick-Anderson commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,178 @@
+
+The following is not a complete list of errors. Should you encounter an error not listed here, please leave a detailed error message in the DISQUS section below (click `Show comments` to open the DISQUS panel).
+
+To diagnose problems with IIS deployments, study browser output, examine the server's **Application** log through **Event Viewer**, and `enable logging <https://docs.asp.net/en/latest/fundamentals/logging.html>`__ in the application. The **ASP.NET Core Module** log will be found on the path provided in the `stdoutLogFile` attribute of the `\<aspNetCore\>` element in `web.config`. Any folders on t ...
```
> to create the `web.config` file will default the `stdoutLogEnabled` to `false`, 
 to create the *web.config* file will default the `stdoutLogEnabled` set to `false`, 
added set

GuardRex replied:
"set" reads strange to me, because it was interpreted as "collection of things" when I read it. How about "setting?"

> Applications that use the `dotnet-publish-iis` tooling to create the `web.config` file will default the `stdoutLogEnabled` setting to `false`, so you must ...

Also, @moozzyk is it "dotnet-publish-iis" tooling or "publish-iis" tooling?

Rick-Anderson replied:
@GuardRex  great

GuardRex replied:
@moozzyk Nevermind ... I see you have it as "publish-iis" tool over in the repo. I'm good. :+1:

moozzyk replied:
the name of the binary is `dotnet-publish-iis` since all dotnet commands have to start with the `dotnet-` prefix. However the tool is commonly invoked as `dotnet publihsh-iis` so I think `publish-iis` is fine.

Rick-Anderson commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,178 @@
+To diagnose problems with IIS deployments, study browser output, examine the server's **Application** log through **Event Viewer**, and `enable logging <https://docs.asp.net/en/latest/fundamentals/logging.html>`__ in the application. The **ASP.NET Core Module** log will be found on the path provided in the `stdoutLogFile` attribute of the `\<aspNetCore\>` element in `web.config`. Any folders on t ...
+
+A quick way to determine if the IIS reverse proxy to the Kestrel server is working properly is to perform a simple static file request (e.g., a stylesheet, script, or image) from the application's static assets (i.e., those in `wwwroot`) using `Static File Middleware <https://docs.asp.net/en/latest/fundamentals/static-files.html>`__. If the application can serve static files but MVC Views and oth ...
+
```
@shirhatti  This is my favorite approach to troubleshooting MVC apps problems on IIS

1. Create a new MVC app without individual user account.
2. Deploy and test.
3. Add complexity to MVC test app.
4. Remove complexity from broken app.
5. Repeat 3-4 until you solve the problem.

That's too long winded. Maybe just tell them to deploy something very simple, such as a new MVC app without individual user accounts.

GuardRex replied:
The reason I included this is that it quickly assesses the status of the IIS-module leg of the config. I use this test quite a lot when I'm helping devs in a chat situation (e.g., JabbR), and it quickly puts us in the right direction to go with troubleshooting.

> tell them to deploy something very simple

This is also very good as a general suggestion, but I'm not sure how to work it in as general advice into a "publishing to IIS" doc. I have told devs who show up with very broken projects to take a step back with a cli-sample or IISIntegration sample and get that running, then go forward with their mega-monster app.

shirhatti replied:
I'm with @GuardRex on this one.

Also, I hope publishing to sample isn't that hard that people have to resort to publishing a placeholder app before they deploy their real application 😄

Rick-Anderson replied:
OK, skip that. It's probably pretty obvious.

Rick-Anderson commented:
Looks good. A few minor tweaks.

GuardRex commented:
@Rick-Anderson Thanks. I gotcha on the changes needed. I'm going to hit this tomorrow (Friday) and will have updates committed by mid-afternoon CST. This will also give @shirhatti some time to get back on points you raised.

GuardRex commented:
@Rick-Anderson @shirhatti Ok ... I think I hit everything, and I found a few other tiny problems.

danroth27 commented on aspnet/publishing/iis.rst:
```diff
@@ -1,116 +1,182 @@
-#. Navigate to the publish folder (*..\\..\\artifacts\\bin\\WebApp9\\Release\\Publish folder* in this sample).
-#. Copy the **approot** and **wwwroot** directories to the target IIS server. Note: MSDeploy is the recommended mechanism for deployment, but you can use Xcopy, Robocopy or another approach. For information on using `Web Deploy` see :doc:`iis-with-msdeploy`.
-#. In IIS Manager, create a new web site and set the physical path to **wwwroot**. You can click on **Browse *.80(http)** to see your deployed app in the browser.
+#. Install the `.NET Core Windows Server Hosting <https://microsoft.com/net>`__ bundle on the server. The bundle will install the .NET Core Runtime, .NET Core Library, and the ASP.NET Core Module. The module creates the reverse-proxy between IIS and the Kestrel server.
```
> Install the `.NET Core Windows Server Hosting <https://microsoft.com/net>`__ 

Install the `.NET Core Windows Server Hosting <http://go.microsoft.com/fwlink/?LinkId=798480>`__

danroth27 commented:
:shipit:

danroth27 commented:
Thanks! e107475f1d522560566ed4e91746ce66d9935b83
